plugin/pkg/admission/jsonpath/admission.go

```diff
@@ -0,0 +1,220 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package jsonpath
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"regexp"
+
+	clientset "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
+
+	"k8s.io/kubernetes/pkg/admission"
+	"k8s.io/kubernetes/pkg/api"
+	util_jsonpath "k8s.io/kubernetes/pkg/util/jsonpath"
+	"k8s.io/kubernetes/pkg/util/yaml"
+)
+
+var jsonRegexp = regexp.MustCompile("^\\{\\.?([^{}]+)\\}$|^\\.?([^{}]+)$")
+
+// MassageJSONPath attempts to be flexible with JSONPath expressions, it accepts:
+//   * metadata.namespace (no leading '.' or curly brances '{...}'
+//   * {metadata.namespace} (no leading '.')
+//   * .metadata.namespace (no curly braces '{...}')
+//   * {.metadata.namespace} (complete expression)
+// And transforms them all into a valid jsonpat expression:
+//   {.metadata.namespace}
+// TODO: This is copied from pkg/kubectl/custom_column_printer.go move into the pkg/util/jsonpath package.
+func massageJSONPath(pathExpression string) (string, error) {
+	if len(pathExpression) == 0 {
+		return pathExpression, nil
+	}
+	submatches := jsonRegexp.FindStringSubmatch(pathExpression)
+	if submatches == nil {
+		return "", fmt.Errorf("unexpected path string, expected a 'name1.name2' or '.name1.name2' or '{name1.name2}' or '{.name1.name2}'")
+	}
+	if len(submatches) != 3 {
+		return "", fmt.Errorf("unexpected submatch list: %v", submatches)
+	}
+	var fieldSpec string
+	if len(submatches[1]) != 0 {
+		fieldSpec = submatches[1]
+	} else {
+		fieldSpec = submatches[2]
+	}
+	return fmt.Sprintf("{.%s}", fieldSpec), nil
+}
+
+type JSONPathAdmissionRule struct {
+	// Version indicates the version for the rule.  For now this has to be "v1"
+	Version string `json:"version" yaml:"version"`
+
+	// Name is the name of this rule
+	Name string `json:"name" yaml:"name"`
+
+	// KindRegexp holds a regular expression used to determine which Kinds of
+	// objects this rule applies to.
+	KindRegexp string `json:"kindRegexp" yaml:"kindRegexp"`
+
+	// Path holds a JSONPath expression to select particular fields
+	Path string `json:"path" yaml:"path"`
+
+	// APIVersion holds the version of the object to use with the
+	// JSONPath above.  (e.g. "v1" or "v1beta1")
+	APIVersion string `json:"apiVersion" yaml:"apiVersion"`
+
+	// MatchRegexp holds a matching expression, all values returned by the JSONPath expression (above)
+	// must match this expression for the object to be admitted
+	MatchRegexp string `json:"matchRegexp" yaml:"matchRegexp"`
+
+	// AcceptEmptyMatch indicates if an empty match is acceptable.  Empty match means that the
+	// JSONPath expression above returns no values.  Default is false, meaning that at least one
+	// value that matches MatchRegexp is required for admission
+	AcceptEmptyMatch bool `json:"acceptEmptyMatch" yaml:"acceptEmptyMatch"`
+}
+
+type JSONPathAdmissionConfig struct {
```

Comment:
this object should have type meta at the very least, would be even better set up like a component config object.